File: hex_game_project_02122/src/main/java/dk/dtu/computer_opponent/MCTSNode.java
package dk.dtu.computer_opponent;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Random;

import dk.dtu.main.Coordinate;
import dk.dtu.main.GameBoard;

public class MCTSNode {
    private static int counter = 0;
    private Coordinate move;
    private MCTSNode parent;
    private List<MCTSNode> children;
    private boolean expanded;

    private int visits;
    private double wins;

    private int playerNumber;
    private static final double EXPLORATION_PARAMETER = 0.7;
    private static final double EPSILON = 1e-6;
    private Random rand;

    private static long totalSelectionTime = 0;
    private static long totalExpansionTime = 0;
    private static long totalSimulationTime = 0;
    private static long totalBackpropagationTime = 0;
    private static int runCount = 0;

    public MCTSNode(MCTSNode parent, Coordinate move, int playerNumber) {
        this.move = move;
        this.parent = parent;
        this.playerNumber = playerNumber;
        this.children = new ArrayList<>();
        this.visits = 0;
        this.wins = 0;
        this.expanded = false;
        this.rand = new Random();
    }

    public void selectAction(GameBoard gameBoard) {
        long startTime, endTime;

        List<MCTSNode> visited = new LinkedList<>();
        MCTSNode current = this;
        visited.add(current);

        startTime = System.nanoTime();
        while (!current.isLeaf()) {
            current = current.selection();
            visited.add(current);
        }
        endTime = System.nanoTime();
        totalSelectionTime += (endTime - startTime);

        if (!current.expanded) {
            startTime = System.nanoTime();
            current.expansion(gameBoard);
            current.expanded = true;
            endTime = System.nanoTime();
            totalExpansionTime += (endTime - startTime);
        }

        startTime = System.nanoTime();
        MCTSNode nodeToRollout = current;
        double value = simulation(nodeToRollout, gameBoard);
        endTime = System.nanoTime();
        totalSimulationTime += (endTime - startTime);

        startTime = System.nanoTime();
        for (MCTSNode node : visited) {
            node.backpropagation(value);
        }
        endTime = System.nanoTime();
        totalBackpropagationTime += (endTime - startTime);

        runCount++;
    }

    public static void printTimingStats() {
        System.out.println("After " + runCount + " runs:");
        System.out.println("Avg Selection Time:      " + (totalSelectionTime / runCount) / 1_000.0 + " µs");
        System.out.println("Avg Expansion Time:      " + (totalExpansionTime / runCount) / 1_000.0 + " µs");
        System.out.println("Avg Simulation Time:     " + (totalSimulationTime / runCount) / 1_000.0 + " µs");
        System.out.println("Avg Backpropagation Time:" + (totalBackpropagationTime / runCount) / 1_000.0 + " µs");
        long total = totalSelectionTime + totalExpansionTime + totalSimulationTime + totalBackpropagationTime;
        System.out.println("Avg Total Time Per Iteration: " + (double)(total / runCount) / 1_000.0 + " µs\n");
    }

    public boolean isLeaf() {
        return children.isEmpty();
    }

    public void expansion(GameBoard gameBoard) {
        List<Coordinate> availableMoves = gameBoard.getAvailableMoves();

        int nextPlayer = (playerNumber == 1) ? 2 : 1;
        for (Coordinate availableMove : availableMoves) {
            MCTSNode child = new MCTSNode(this, availableMove, nextPlayer);
            children.add(child);
        }
    }

    private MCTSNode selection() {
        MCTSNode selected = null;
        double bestValue = Double.NEGATIVE_INFINITY;

        for (MCTSNode child : children) {
            double exploitationTerm = child.wins / (child.visits + EPSILON);
            double explorationTerm = EXPLORATION_PARAMETER
                    * Math.sqrt(Math.log(this.visits + 1) / (child.visits + EPSILON));
            double uctScore = exploitationTerm + explorationTerm;

            if (uctScore > bestValue) {
                selected = child;
                bestValue = uctScore;
            }
        }

        return selected;
    }

    private double simulation(MCTSNode node, GameBoard gameBoard) {
        SimulationGame simGame = new SimulationGame(gameBoard);

        applyMovesToSimulation(node, simGame);

        int currentPlayer = node.playerNumber;

        List<Coordinate> availableMoves = simGame.getAvailableMoves();
        int totalMoves = availableMoves.size();
        int moveCount = 0;

        while (moveCount < totalMoves) {
            int i = rand.nextInt(availableMoves.size());
            Coordinate spot = availableMoves.get(i);
            simGame.makeMove(spot, currentPlayer);

            availableMoves.remove(spot);

            moveCount++;
            currentPlayer = (currentPlayer == 1) ? 2 : 1;
        }
        simGame.checkWin();

        return (simGame.winner == this.playerNumber) ? 1.0 : 0.0;
    }

    private void applyMovesToSimulation(MCTSNode node, SimulationGame simGame) {
        List<MCTSNode> pathToRoot = new ArrayList<>();
        MCTSNode current = node;

        while (current.parent != null) {
            pathToRoot.add(0, current);
            current = current.parent;
        }

        for (MCTSNode pathNode : pathToRoot) {
            if (pathNode.move != null && pathNode.parent != null) {
                Coordinate move = pathNode.move;
                int player = pathNode.parent.playerNumber;
                simGame.board[move.getX()][move.getY()].setState(player);
            }
        }
    }

    public void backpropagation(double value) {
        visits++;
        wins += value;
    }

    public boolean isFullyExpanded() {
        return expanded;
    }

    public int getVisits() {
        return visits;
    }

    public double getWins() {
        return wins;
    }

    public int getPlayerNumber() {
        return playerNumber;
    }

    public List<MCTSNode> getChildren() {
        return children;
    }

    public Coordinate getMove() {
        return move;
    }

    public void addChild(MCTSNode child) {
        children.add(child);
    }
}

File: hex_game_project_02122/native/src/MCTSNode.cpp
#include "MCTSNode.h"
#include "MCTSUtils.h"
#include <iostream>
#include <chrono>
#include <iostream>

static long long totalSelectionTime = 0;
static long long totalExpansionTime = 0;
static long long totalSimulationTime = 0;
static long long totalBackpropagationTime = 0;
static int runCount = 0;


std::mt19937 MCTSNode::shared_rng(std::random_device{}());

MCTSNode::MCTSNode(MCTSNode* parent, Move move, int playerNumber)
    : parent(parent),
      move(move),
      playerNumber(playerNumber),
      expanded(false),
      visits(0),
      wins(0) {}


void MCTSNode::selectAction(const std::vector<std::vector<int>>& board) {
    using namespace std::chrono;

    auto start = high_resolution_clock::now();

    std::vector<MCTSNode*> visited;
    MCTSNode* current = this;
    visited.push_back(current);

    // --- Selection ---
    auto selStart = high_resolution_clock::now();
    while (!current->isLeaf()) {
        current = current->selection();
        visited.push_back(current);
    }
    auto selEnd = high_resolution_clock::now();
    totalSelectionTime += duration_cast<microseconds>(selEnd - selStart).count();

    // --- Expansion ---
    if (!current->expanded) {
        auto expStart = high_resolution_clock::now();
        current->expansion(board);
        current->expanded = true;
        auto expEnd = high_resolution_clock::now();
        totalExpansionTime += duration_cast<microseconds>(expEnd - expStart).count();
    }

    // --- Simulation ---
    auto simStart = high_resolution_clock::now();
    MCTSNode* nodeToRollout = current;
    double value = simulation(nodeToRollout, board);
    auto simEnd = high_resolution_clock::now();
    totalSimulationTime += duration_cast<microseconds>(simEnd - simStart).count();

    // --- Backpropagation ---
    auto backStart = high_resolution_clock::now();
    for (MCTSNode* node : visited) {
        node->backPropagation(value);
    }
    auto backEnd = high_resolution_clock::now();
    totalBackpropagationTime += duration_cast<microseconds>(backEnd - backStart).count();

    runCount++;
}

void MCTSNode::printTimingStats() {
    std::cout << "After " << runCount << " runs (times in µs):\n";
    std::cout << "Avg Selection Time:       " << (double)totalSelectionTime / runCount << " µs\n";
    std::cout << "Avg Expansion Time:       " << (double)totalExpansionTime / runCount << " µs\n";
    std::cout << "Avg Simulation Time:      " << (double)totalSimulationTime / runCount << " µs\n";
    std::cout << "Avg Backpropagation Time: " << (double)totalBackpropagationTime / runCount << " µs\n";

    long long total = totalSelectionTime + totalExpansionTime + totalSimulationTime + totalBackpropagationTime;
    std::cout << "Avg Total Time Per Iteration: " << (double)total / runCount << " µs\n";
}

bool MCTSNode::isLeaf() {
    return children.empty();
}

void MCTSNode::expansion(const std::vector<std::vector<int>>& board) {
    const std::vector<Move>& availableMoves = getAvailableMoves(board);  // Avoid copy
    children.reserve(availableMoves.size());

    int nextPlayer = playerNumber == 1 ? 2 : 1;
    for (const auto& move : availableMoves) {
        children.push_back(new MCTSNode(this, move, nextPlayer));
    }
}

MCTSNode* MCTSNode::selection() {
    MCTSNode* selected = nullptr;
    double bestValue = -std::numeric_limits<double>::infinity();

    const double logVisits = std::log(static_cast<double>(this->visits + 1));

    std::uniform_real_distribution<double> noiseDist(0.0, EPSILON);

    for (MCTSNode* child : children) {
        double childVisits = static_cast<double>(child->visits) + EPSILON;
        double exploitationTerm = child->wins / childVisits;

        double explorationTerm = EXPLORATION_PARAMETER * std::sqrt(logVisits / childVisits);
        explorationTerm += noiseDist(shared_rng);

        double uctScore = exploitationTerm + explorationTerm;

        if (uctScore > bestValue) {
            bestValue = uctScore;
            selected = child;
        }
    }

    return selected;
}

double MCTSNode::simulation(MCTSNode* node, const std::vector<std::vector<int>>& originalBoard) {
    std::vector<MCTSNode*> pathToRoot;
    MCTSNode* current = node;
    std::vector<std::vector<int>> board = originalBoard;

    while (current->parent != nullptr) {
        pathToRoot.push_back(current);
        current = current->parent;
    }
    std::reverse(pathToRoot.begin(), pathToRoot.end());

    for (MCTSNode* pathNode : pathToRoot) {
        if (!(pathNode->move == Move(-1, -1)) && pathNode->parent != nullptr) {
            int x = pathNode->move.x;
            int y = pathNode->move.y;
            int player = pathNode->parent->playerNumber;
            board[x][y] = player;
        }
    }

    int currentPlayer = node->playerNumber;
    int winner = 0;

    std::vector<Move> availableMoves = getAvailableMoves(board);
    int totalMoves = availableMoves.size();
    int moveCount = 0;

    while (moveCount < totalMoves) {
        std::uniform_int_distribution<int> dist(0, availableMoves.size() - 1);
        auto move = availableMoves[dist(MCTSNode::shared_rng)];
        board[move.x][move.y] = currentPlayer;

        auto it = std::find(availableMoves.begin(), availableMoves.end(), move);
        if (it != availableMoves.end()) {
            availableMoves.erase(it);
        }
        
        moveCount++;
        currentPlayer = (currentPlayer == 1) ? 2 : 1;
    }
    
    for (int i = 0; i < board.size(); ++i) {
        if (board[i][0] == 2 && checkWinningMove(Move(i, 0), board, 2)) {
            winner = 2;
            break;
        }
        winner = 1;
    }

    return (winner == this->playerNumber) ? 1.0 : 0.0;
}

void MCTSNode::backPropagation(double value) {
    visits++;
    wins += value;
}

MCTSNode* MCTSNode::getParent() {
    return parent;
}

Move MCTSNode::getMove() {
    return move;
}

int MCTSNode::getPlayerNumber() {
    return playerNumber;
}

const std::vector<MCTSNode*>& MCTSNode::getChildren() {
    return children;
}

int MCTSNode::getVisits() {
    return visits;
}

MCTSNode::~MCTSNode() {
    for (MCTSNode* child : children) {
        delete child;
    }
}

